Ldap style oracle jdbc urls not supported for native

### Describe the bug

If I use a url `jdbc:oracle:thin:@ldap://...`, I get the following error at startup: ` java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory]`



### Expected behavior

Should work same as non ldap oracle jdbc url.

### Actual behavior

Here is the full exception:

```
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
2021-07-28 11:51:47,444 WARN  [io.agr.pool] (agroal-11) Datasource '<default>': IO Error: JNDI Package failure javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory] (CONNECTION_ID=yh0svrfuQK+prg+PVtb/mw==)
2021-07-28 11:51:47,445 WARN  [org.hib.eng.jdb.env.int.JdbcEnvironmentInitiator] (JPA Startup Thread: <default>) HHH000342: Could not obtain connection to query metadata: java.sql.SQLRecoverableException: IO Error: JNDI Package failure javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory] (CONNECTION_ID=yh0svrfuQK+prg+PVtb/mw==)
        at oracle.jdbc.driver.T4CConnection.handleLogonNetException(T4CConnection.java:870)
        at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:675)
        at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:1032)
        at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:90)
        at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:681)
        at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:602)
        at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:204)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:470)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:452)
        at java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:68)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(Thread.java:829)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: oracle.net.ns.NetException: JNDI Package failure javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory] (CONNECTION_ID=yh0svrfuQK+prg+PVtb/mw==)
        at oracle.net.jndi.JndiAttrs.initializeLDAPContext(JndiAttrs.java:290)
        at oracle.net.jndi.JndiAttrs.<init>(JndiAttrs.java:165)
        at oracle.net.resolver.AddrResolution.<init>(AddrResolution.java:334)
        at oracle.net.ns.NSProtocol.connect(NSProtocol.java:329)
        at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1967)
        at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:640)
        ... 14 more
Caused by: javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory]
        at javax.naming.spi.NamingManager.getFactory(NamingManager.java:760)
        at javax.naming.spi.NamingManager.lambda$getInitialContext$1(NamingManager.java:722)
        at jdk.internal.loader.AbstractClassLoaderValue$Memoizer.get(AbstractClassLoaderValue.java:329)
        at jdk.internal.loader.AbstractClassLoaderValue.computeIfAbsent(AbstractClassLoaderValue.java:205)
        at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:722)
        at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
        at javax.naming.InitialContext.init(InitialContext.java:236)
        at javax.naming.InitialContext.<init>(InitialContext.java:208)
        at javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)
        at oracle.net.jndi.JndiAttrs.initializeLDAPContext(JndiAttrs.java:282)
        ... 19 more
Caused by: java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory
        at com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:64)
        at java.lang.Class.forName(DynamicHub.java:1308)
        at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:127)
        at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:133)
        at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:96)
        at javax.naming.spi.NamingManager.getFactory(NamingManager.java:757)
        ... 28 more
```

### How to Reproduce?

Create an application using the `quarkus-jdbc-oracle` driver.
Compile the application in native.
Run the application with a url of type: `quarkus.datasource.jdbc.url=jdbc:oracle:thin:@ldap://...`.
The application fails at startup.



### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

GraalVM 21.1.0 Java 11 CE (Java Version 11.0.11+8-jvmci-21.1-b05)

### Quarkus version or git rev

2.1.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Note that the application works with tns style url: `quarkus.datasource.jdbc.url=jdbc:oracle:thin:@(DESCRIPTION=(ENABLE=BROKEN)(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)...` (the address resolved by ldap).
So the issue is really about ldap access, not oracle.

@gsmet I have a similar issue with the following trace:
```
Caused by: java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory
at com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:64)
at java.lang.Class.forName(DynamicHub.java:1308)
at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:127)
at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:133)
at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:96)
at javax.naming.spi.NamingManager.getFactory(NamingManager.java:757)
... 28 more
```

Can someone knowledgeable have a look, please?

--- a/extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java
+++ b/extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java
@@ -21,6 +21,9 @@ public final class OracleReflections {
         // but it delegates all use to "oracle.jdbc.driver.OracleDriver" - which is also what's recommended by the docs.
         final String driverName = "oracle.jdbc.driver.OracleDriver";
         reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, driverName));
+
+        // for ldap style jdbc urls. e.g. jdbc:oracle:thin:@ldap://oid:5000/mydb1,cn=OracleContext,dc=myco,dc=com
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.jndi.ldap.LdapCtxFactory"));
     }
 
 }

--- a/integration-tests/jpa-oracle/src/main/java/io/quarkus/example/jpaoracle/LdapUrlTestEndpoint.java
+++ b/integration-tests/jpa-oracle/src/main/java/io/quarkus/example/jpaoracle/LdapUrlTestEndpoint.java
@@ -1,0 +1,52 @@
+package io.quarkus.example.jpaoracle;
+
+import java.io.IOException;
+import java.sql.SQLException;
+
+import javax.inject.Inject;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.jboss.logging.Logger;
+
+import io.agroal.api.AgroalDataSource;
+import io.quarkus.agroal.DataSource;
+
+@WebServlet(name = "JPATestOracleLdap", urlPatterns = "/jpa-oracle/testldap")
+public class LdapUrlTestEndpoint extends HttpServlet {
+
+    private final Logger LOG = Logger.getLogger(LdapUrlTestEndpoint.class.getName());
+
+    @Inject
+    @DataSource("ldap")
+    AgroalDataSource ds;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        try {
+            final String output = ldap();
+            resp.getWriter().write(output);
+
+        } catch (Exception e) {
+            resp.getWriter().write("An error occurred while attempting ldap operations");
+        }
+    }
+
+    private String ldap() throws SQLException {
+
+        try {
+            ds.getConnection().close();
+        } catch (SQLException e) {
+            LOG.info("received exception: " + e);
+            if (e.toString().contains("java.net.UnknownHostException: oid")) {
+                return "OK";
+            }
+            throw e;
+        }
+
+        return "KO: did not get expected exception";
+    }
+
+}

--- a/integration-tests/jpa-oracle/src/main/resources/application.properties
+++ b/integration-tests/jpa-oracle/src/main/resources/application.properties
@@ -5,3 +5,8 @@ quarkus.datasource.jdbc.url=${oracledb.url}
 quarkus.datasource.jdbc.max-size=1
 quarkus.hibernate-orm.database.generation=drop-and-create
 quarkus.datasource.jdbc.additional-jdbc-properties."oracle.jdbc.timezoneAsRegion"=false
+
+quarkus.datasource.ldap.db-kind=oracle
+quarkus.datasource.ldap.username=SYSTEM
+quarkus.datasource.ldap.password=hibernate_orm_test
+quarkus.datasource.ldap.jdbc.url=jdbc:oracle:thin:@ldap://oid:5000/mydb1,cn=OracleContext,dc=myco,dc=com

--- a/integration-tests/jpa-oracle/src/test/java/io/quarkus/it/jpa/oracle/JPAFunctionalityTest.java
+++ b/integration-tests/jpa-oracle/src/test/java/io/quarkus/it/jpa/oracle/JPAFunctionalityTest.java
@@ -25,4 +25,9 @@ public class JPAFunctionalityTest {
         RestAssured.when().get("/jpa-oracle/testserialization").then().body(is("Hello from Serialization Test"));
     }
 
+    @Test
+    public void testLdapFromServlet() throws Exception {
+        RestAssured.when().get("/jpa-oracle/testldap").then().body(is("OK"));
+    }
+
 }

